doc/design/migration_to_python.md

```diff
@@ -0,0 +1,479 @@
+
+## A Quick View Of The Current Architecture
+
+Before diving into the migration, we should walk through the current packages to judiciously judge what would migrate to python.
+
+### The Current Python Package
+
+The main python *ML code* are in the python package [sqlflow_submitter](https://github.com/sql-machine-learning/sqlflow/tree/400c691470c6503393453d47856913df3365503e/python/sqlflow_submitter). The main purpose of the golang *ML code* is to generate python scripts that call python functions in the `sqlflow_submitter`  package. We call these python scripts the *submitter programs*. Because `sqlflow_submitter` is already written in python. There will be little modification to this package in the migration process.
+
+### The Current Golang Package Structure
+
+At the moment, the packages under [sqlflow/pkg](https://github.com/sql-machine-learning/sqlflow/tree/400c691470c6503393453d47856913df3365503e/pkg) are:
+
+1. [server](https://github.com/sql-machine-learning/sqlflow/tree/400c691470c6503393453d47856913df3365503e/pkg/server) is the implementation of the SQLFlow gRPC server. The input statements from the user will be sent to the SQLFlow gRPC server. It forwards the statements to the `workflow` package and get results back. We **don't** have to migrate this package to python.
+
+2. [proto](https://github.com/sql-machine-learning/sqlflow/tree/400c691470c6503393453d47856913df3365503e/pkg/proto)  defines necessary pb messages for the `server` package. We **don't** have to migrate this package to python.
+
+3. [workflow](https://github.com/sql-machine-learning/sqlflow/tree/400c691470c6503393453d47856913df3365503e/pkg/workflow) is the downstream of the `server` package. It does a text substitution and generates a python script. The python script calls [couler](https://github.com/sql-machine-learning/sqlflow/tree/400c691470c6503393453d47856913df3365503e/doc/design/couler_sqlflow.md) to generate an `Argo`  `.yaml` file that defines the workflow to be executed on Argo. We propose to migrate this package to python for two reasons:
+
+   1. The `workflow`  package has the same design as the *ML code* described above, it has similar flaws. 
+   2. In the next version of the `COLUMN` clause, the `COLUMN` clause may generate its own workflow steps to derive features or get analysis data. This process will requires some *ML code* and should be implemented in python.
+
+4. [step](https://github.com/sql-machine-learning/sqlflow/tree/400c691470c6503393453d47856913df3365503e/pkg/step) is the downstream of the `workflow` package, it's binary `cmd/step` is executed as a workflow step on Argo. The `step/feature` directory should be migrated to python because it has *ML code* that deduces feature columns from the input. 
```

Comment:
Is a step's input still a SQL statement after migration? And the step parse and execute the SQL?

Reply:
Yes, see earlier discussion https://github.com/sql-machine-learning/sqlflow/pull/2335#discussion_r431035131